pkg/controlplane/control/peer.go

```diff
@@ -0,0 +1,276 @@
+	"github.com/clusterlink-net/clusterlink/pkg/util/tls"
+)
+
+const (
+	heartbeatInterval = 10 * time.Second
+)
+
+// peerMonitor monitors a single peer.
+type peerMonitor struct {
+	lock           sync.Mutex
+	pr             *v1alpha1.Peer
+	client         *peer.Client
+	statusCallback func(*v1alpha1.Peer)
+
+	wg     *sync.WaitGroup
+	stopCh chan struct{}
+
+	logger *logrus.Entry
+}
+
+// peerManager manages peers status.
+type peerManager struct {
+	client         client.Client
+	peerTLS        *tls.ParsedCertData
+	statusCallback func(*v1alpha1.Peer)
+
+	lock     sync.Mutex
+	monitors map[string]*peerMonitor
+
+	stopped         bool
+	monitorWG       sync.WaitGroup
+	updaterWG       sync.WaitGroup
+	stopCh          chan struct{}
+	statusUpdatesCh chan *v1alpha1.Peer
+
+	logger *logrus.Entry
+}
+
+func (m *peerMonitor) Peer() v1alpha1.Peer {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return *m.pr
+}
+
+func (m *peerMonitor) Start() {
+	defer m.wg.Done()
+
+	ticker := time.NewTicker(heartbeatInterval)
+	defer ticker.Stop()
+
+	backoffConfig := backoff.NewExponentialBackOff()
+
+	reachable := false
+	reachableCond := metav1.Condition{
+		Type:   v1alpha1.PeerReachable,
+		Status: metav1.ConditionFalse,
+		Reason: "Heartbeat",
+	}
+
+	for {
+		select {
+		case <-m.stopCh:
+			return
+		default:
+			break
+		}
+
+		err := backoff.Retry(m.client.GetHeartbeat, backoffConfig)
+		if heartbeatOK := err == nil; heartbeatOK != reachable {
+			m.logger.Infof("Heartbeat result: %v", heartbeatOK)
+
+			if heartbeatOK {
+				reachableCond.Status = metav1.ConditionTrue
+				backoffConfig.MaxElapsedTime = heartbeatInterval
+			} else {
+				reachableCond.Status = metav1.ConditionFalse
+				backoffConfig.MaxElapsedTime = 0
+			}
+
+			reachable = heartbeatOK
+
+			m.lock.Lock()
+			meta.SetStatusCondition(&m.pr.Status.Conditions, reachableCond)
+			m.lock.Unlock()
+
+			m.statusCallback(m.pr)
+		}
+
+		// wait till it's time for next heartbeat round
+		<-ticker.C
+	}
+}
+
+func (m *peerMonitor) Stop() {
+	close(m.stopCh)
+}
+
+// AddPeer defines a new route target for egress dataplane connections.
+func (m *peerManager) AddPeer(pr *v1alpha1.Peer) {
+	m.logger.Infof("Adding peer '%s'.", pr.Name)
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.stopped {
+		return
+	}
+
+	monitor, ok := m.monitors[pr.Name]
+	if !ok || peerChanged(monitor.pr, pr) {
+		m.monitors[pr.Name] = newPeerMonitor(pr, m)
+	}
+}
+
+// DeletePeer removes the possibility for egress dataplane connections to be routed to a given peer.
+func (m *peerManager) DeletePeer(name string) {
+	m.logger.Infof("Deleting peer '%s'.", name)
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	delete(m.monitors, name)
+}
+
+// Name of the peer monitor runnable.
+func (m *peerManager) Name() string {
+	return "peerManager"
+}
+
+func (m *peerManager) SetStatusCallback(callback func(*v1alpha1.Peer)) {
+	m.statusCallback = callback
+}
+
+// Start the peer manager.
+func (m *peerManager) Start() error {
```

Comment:
Where do you call this function?

Reply:
It's called by the runnable manager, which subscribes here:
https://github.com/clusterlink-net/clusterlink/blob/b41c7721d4260568a40aa1f79fc03e26859f90e8/cmd/cl-controlplane/app/server.go#L205